### ISSUE TITLE
dynamic mapping

### ISSUE THREAD

**Issue by peterkracik** [dustin10/VichUploaderBundle#1027]
|    Q         |   A
|--------- | -----
| Version  | 1.10.1

### Support Question

Is there a possibility to have dynamic mapping?
to explain, I've created a class Image, which is used in other classes like: Product (for product image) and Company (company logo). and I would like to set mapping based on the class where it's mapped.

Image.php
```
/**
 * @MongoDB\EmbeddedDocument
 * @MongoDB\HasLifecycleCallbacks
 * @Vich\Uploadable
 */
class Image
{
    /**
     * @MongoDB\Id()
     */
    protected $id;

    /**
     * @Vich\UploadableField(**mapping="product",** fileNameProperty="imageName", size="imageSize")
     * @var File
     */
    private $imageFile;

     /**
     * @MongoDB\Field(type="string")
     */
    private $imageName;

    /**
     * @MongoDB\Field(type="integer")
     */
    private $imageSize;

    /**
     * @MongoDB\Field(type="date")
     */
    private $updatedAt;

    private $finalName;
```

and then classes like Company.php
```
/**
 * @MongoDB\Document
 */
class Company
{
    /**
     * @MongoDB\EmbedOne(targetDocument="Image")
     * @var Image
     */
    protected $logo;
 ...
```

and Product
```
/**
 * @MongoDB\Document
 */
class Product
{
    /**
     * @MongoDB\EmbedOne(targetDocument="Image")
     * @var Image
     */
    protected $image;
 ...
```

and I would like different mapping for these two classes, or at least to have a possibility to use different folder.
is there a way to achieve this?

thank you for the help

**Comment by garak** [comment]
Just define your mappings. I can't see any reason to be "dynamic" here

**Comment by hatto** [comment]
thank you @garak for your fast reply.

the thing is that I like to have a different mapping based on the class where it is used. mapping "product" for image in product class and mapping "logo" for image in company class.
because I want to avoid create child classes like  ProductImage, CompanyLogoImage etc just to have a different folder (mapping).

**Comment by garak** [comment]
I see.
Try to define your mapping on `$image` property of Product and Company and remove mapping from  `$imageFile` in Image.
Something like:

>@Vich\UploadableField(**mapping="product",** fileNameProperty="image.imageName")

**Comment by hatto** [comment]
no, unfortunately it doesnt work. I have error 500 - Mapping not found for field "imageFile"

or maybe some other way how to set mapping? for example in the form? (i use sonata admin), or I tried to create a custom annotation which property I could get inside the image class, I didnt success but maybe it's a way to do it? 
i've spent all day yesterday to find a way how to do it properly, but couldn't figure it out and I suppose I am not the first one with this problem :)

**Comment by garak** [comment]
Your mapping is not on `imageFile` anymore, it's on `image`

**Comment by hatto** [comment]
yes, its what I did

Image.php
```
/**
 * @MongoDB\EmbeddedDocument
 * @MongoDB\HasLifecycleCallbacks
 * @Vich\Uploadable
 */
class Image
{
    /**
     * @MongoDB\Id()
     */
    protected $id;

    /**
     * @var File
     */
    private $imageFile;

     /**
     * @MongoDB\Field(type="string")
     */
    private $imageName;

    /**
     * @MongoDB\Field(type="integer")
     */
    private $imageSize;

    /**
     * @MongoDB\Field(type="date")
     */
    private $updatedAt;

    private $finalName;
```

Product.php
```
/**
 * @MongoDB\Document
 */
class Product
{
    /**
     * @MongoDB\EmbedOne(targetDocument="Image")
     * @Vich\UploadableField(mapping="product", fileNameProperty="imageName", size="imageSize")
     * @var Image
     */
    protected $image;
 ...
```

and this returns me that error

**Comment by garak** [comment]
`fileNameProperty="image.imageName"` (as stated before)

**Comment by hatto** [comment]
I've tried it but unfortunately it doesnt work, because Image in my Product class is not really uploadable field, only object which one contains uplodable file. and this one needs the mapping in the annotation, not the object in the Product class. if its not there I've got the mentioned error. and the imageFile (within Image class) doesnt 'inherit' the mapping from the image property in the Product class.

Class Product
- id :int
- name :string
- image :Image
...

Image.php
 - id :int
 - imageName :string
  - imageFile :file (uploadableFile (mapping="") )
... 

or maybe I am just trying what's not possible with vichUploader?

**Comment by garak** [comment]
It is possible, I'm using myself this bundle with an embedded value object.

Unfortunately, in my project there's no Mongo (I'm using Doctrine ORM) and there are no annotations (I'm using yaml mapping instead)

**Comment by Akekc** [comment]
Hi @garak  , I using Doctrine and have the same issue. 
Could you please describe how I can solve it?

I want to use 1 Entity (MediaObject ) and relate it with several other Entities (User, Product, etc). But files need to be stored at the different directories ( regarding mapping )

**Comment by garak** [comment]
> Could you please describe how I can solve it?

I didn't solve it. Dynamic mapping is not supported at the moment.

**Comment by Akekc** [comment]
Damn, well, thanks for the answer anyway)

**Comment by lukasz-zaroda** [comment]
Has anything changed? To be fair, it seems to me that almost the same things as dynamic mapping can be achieved by simply providing custom namers, that would dynamically provide names/paths based on some property saved in the entity representing the file.

**Comment by garak** [comment]
> Has anything changed? 

No

**Comment by garak** [comment]
Closing, since this seems a requirement too specific to be directly supported in this bundle.
If you think you can propose a general-purpose solution, feel free to propose a PR.
